src/main/java/uk/gov/ons/census/action/schedule/ActionRuleProcessor.java

```diff
@@ -79,15 +81,17 @@ private void createScheduledActions(ActionRule triggeredActionRule) {
   private void executeAllCases(ActionRule triggeredActionRule) {
     String actionPlanId = triggeredActionRule.getActionPlan().getId().toString();
 
-    try (Stream<Case> cases = caseRepository.findByActionPlanId(actionPlanId)) {
+    Specification<Case> specification = createSpecificationForUnreceiptedCases(actionPlanId);
+
+    try (Stream<Case> cases = caseRepository.findAll(specification).stream()) {
```

Comment:
the `.stream()` function is not the same as the streaming API (i.e. it is not using the same efficient streaming mechanism).

You should use the repo method `findByActionPlanIdAndReceiptReceivedIsFalse` which returns a `Stream`.

Reply:
Done